# The following corespond to the names of DNA Feature Categories
FEATURE_TYPES = (
    # http://www.insdc.org/documents/feature_table.html#7.3.1
    'assembly_gap',
    'attenuator',
    'C_region',
    'CAAT_signal',
    'CDS',
    'centromere',
    'D-loop',
    'D_segment',
    'enhancer',
    'exon',
    'gap',
    'GC_signal',
    'gene',
    'iDNA',
    'intron',
    'J_segment',
    'LTR',
    'mat_peptide',
    'misc_binding',
    'misc_difference',
    'misc_feature',
    'misc_recomb',
    'misc_RNA',
    'misc_signal',
    'misc_structure',
    'mobile_element',
    'modified_base',
    'mRNA',
    'ncRNA',
    'N_region',
    'old_sequence',
    'operon',
    'oriT',
    'polyA_signal',
    'polyA_site',
    'precursor_RNA',
    'prim_transcript',
    'primer_bind',
    'promoter',
    'protein_bind',
    'RBS',
    'repeat_region',
    'rep_origin',
    'rRNA',
    'S_region',
    'sig_peptide',
    'source',
    'stem_loop',
    'STS',
    'TATA_signal',
    'telomere',
    'terminator',
    'tmRNA',
    'transit_peptide',
    'tRNA',
    'unsure',
    'V_region',
    'V_segment',
    'variation',
    "3'UTR",
    "5'UTR",
    '-10_signal',
    '-35_signal',

    # found
    "3'clip",
)

FEATURE_QUALIFIERS = (
    # from http://www.insdc.org/documents/feature_table.html
    '/allele=',
    '/altitude=',
    '/anticodon=',
    '/artificial_location',
    '/bio_material=',
    '/bound_moiety=',
    '/cell_line=',
    '/cell_type=',
    '/chromosome=',
    '/citation=',
    '/clone=',
    '/clone_lib=',
    '/codon_start=',
    '/collected_by=',
    '/collection_date=',
    '/compare=',
    '/country=',
    '/cultivar=',
    '/culture_collection=',
    '/db_xref=',
    '/dev_stage=',
    '/direction=',
    '/EC_number=',
    '/ecotype=',
    '/environmental_sample',
    '/estimated_length=',
    '/exception=',
    '/experiment=',
    '/focus',
    '/frequency=',
    '/function=',
    '/gap_type=',
    '/gene=',
    '/gene_synonym=',
    '/germline',
    '/haplogroup=',
    '/haplotype=',
    '/host=',
    '/identified_by=',
    '/inference=',
    '/isolate=',
    '/isolation_source=',
    '/lab_host=',
    '/lat_lon=',
    '/linkage_evidence=',
    '/locus_tag=',
    '/macronuclear',
    '/map=',
    '/mating_type=',
    '/mobile_element_type=',
    '/mod_base=',
    '/mol_type=',
    '/ncRNA_class=',
    '/note=',
    '/number=',
    '/old_locus_tag=',
    '/operon=',
    '/organelle=',
    '/organism=',
    '/partial',
    '/PCR_conditions=',
    '/PCR_primers=',
    '/phenotype=',
    '/plasmid=',
    '/pop_variant=',
    '/product=',
    '/protein_id=',
    '/proviral',
    '/pseudo',
    '/pseudogene=',
    '/rearranged',
    '/replace=',
    '/ribosomal_slippage',
    '/rpt_family=',
    '/rpt_type=',
    '/rpt_unit_range=',
    '/rpt_unit_seq=',
    '/satellite=',
    '/segment=',
    '/serotype=',
    '/serovar=',
    '/sex=',
    '/specimen_voucher=',
    '/standard_name=',
    '/strain=',
    '/sub_clone=',
    '/sub_species=',
    '/sub_strain=',
    '/tag_peptide=',
    '/tissue_lib=',
    '/tissue_type=',
    '/transgenic',
    '/translation=',
    '/transl_except=',
    '/transl_table=',
    '/trans_splicing',
    '/type_material=',
    '/variety=',

    # compatibility
    '/label=',
)
